### PR TITLE
Add getValueByPath() and setValueByPath()

### PR DESCRIPTION
**Pull request by vjik**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ✔️
| Fixed issues  | #48 

This is break BC. In another packages that use dots-way need replace `getValue()` and `setValue()` to `getValueByPath()` and `setValueByPath()`.

### CODE REVIEW

**Comment by terabytesoftw** [comment]
I am against this PR, since it removes the support for DOT, used by Db and AR, as explained above, in this issue: https://github.com/yiisoft/arrays/issues/23,

**Comment by vjik** [comment]
> am against this PR, since it removes the support for DOT, used by Db and AR, as explained above, in this issue:

replace getValue to getValueByPath

**Comment by rustamwin** [comment]
What are the benefits of having these methods?

**Comment by terabytesoftw** [comment]
Please add tests:

```php
public function testGetValueFromArray($key, $expected, $default = null): void;
public function testGetNestedObjectsValueFromObject(): void;
public function testGetNestedValueFromObjectThatFromArrayFromObject(): void;
public function testGetNestedValueFromObjectFromArray(): void;
public function testGetValueFromStaticProperty(): void
public function testExistingNestedMagicObjectProperty(): void;
public function testNonExistingNestedMagicObjectProperty(): void;
public function testExistingMagicObjectProperty(): void;
public function testNonExistingMagicObjectProperty(): void;
public function testExistingNestedMagicObjectProperty(): void;
public function testNonExistingNestedMagicObjectProperty(): void;
```

I have tested them, they pass all but the first one, those are the AR use cases.

**Comment by vjik** [comment]
> Please add tests:
> 
> ```
> public function testGetValueFromArray($key, $expected, $default = null): void; // no pass tests.
> public function testGetNestedObjectsValueFromObject(): void; // fine
> public function testGetNestedValueFromObjectThatFromArrayFromObject(): void; //fine
> public function testGetNestedValueFromObjectFromArray(): void; //fine
> public function testGetValueFromStaticProperty(): void //fine
> public function testExistingNestedMagicObjectProperty(): void; //fine
> public function testNonExistingNestedMagicObjectProperty(): void; //fine
> public function testExistingMagicObjectProperty(): void; //fine
> public function testNonExistingMagicObjectProperty(): void; //fine
> public function testExistingNestedMagicObjectProperty(): void; //fine
> public function testNonExistingNestedMagicObjectProperty(): void; //fine
> ```
> 
> I have tested them, they pass all but the first one, those are the AR use cases.
> 
> good job.

Tests moved to  `tests/GetValueTest.php`.
Part tests is modified for separately use for getValue and getValueByPath.

File: src/ArrayHelper.php [hunk 1]
```diff
@@ -285,6 +257,120 @@ private static function getRootValue($array, $key, $default)
...
+     *
+     * @param array|object $array array or object to extract value from
+     * @param string|\Closure|array $path key name of the array element, an array of keys or property name
+     * of the object, or an anonymous function returning the value. The anonymous function signature should be:
+     * `function($array, $defaultValue)`.
+     * @param mixed $default the default value to be returned if the specified array key does not exist. Not used when
+     * getting value from an object.
+     * @return mixed the value of the element if found, default value otherwise
+     */
+    public static function getValueByPath($array, $path, $default = null)
```
Comment (terabytesoftw) [inline src/ArrayHelper.php:297]: I would look for a better name path sounds like a directory.

File: src/ArrayHelper.php [hunk 2]
```diff
@@ -337,27 +423,12 @@ private static function getRootValue($array, $key, $default)
      * if the path is null then `$array` will be assigned the `$value`
      * @param mixed $value the value to be written
      */
-    public static function setValue(array &$array, $path, $value): void
+    public static function setValueByPath(array &$array, $path, $value): void
```
Comment (terabytesoftw) [inline src/ArrayHelper.php:426]: I would still look for a better name.

**Review by terabytesoftw — COMMENTED** [review]
Look great 👍

**Comment by vjik** [comment]
Has idea rename methods `getValueByPath` / `setValueByPath`.

Variants:

1. Leave as is: `getValueByPath` / `setValueByPath`.
2. `getValueDots` / `setValueDots`
3. `getValueWithDots` / `setValueWithDots`
4. `getValueUsingDots` / `setValueUsingDots`
5. `getValueByRoute` / `setValueByRoute`

What is your choice?

**Comment by hiqsol** [comment]
Out of these 5 variants I would prefer leave as is.
Also we can explicitly specify path delimiter to make "path" more understandable from method signature:

`getValueByPath($array, $path, $default = null, $pathDelimiter = '.')`
`setValueByPath($array, $path, $value, $pathDelimiter = '.')`

**Comment by smitt14ua** [comment]
If we use suffix "Dots" in method, it will not need $delimiter argument.

**Comment by vjik** [comment]
Added $pathDelimiter to methods getValueByPath and setValueByPath.

**Review by samdark — APPROVED** [review]

**Review by hiqsol — APPROVED** [review]

**Comment by samdark** [comment]
👍

**Pull request merged**